core/src/main/java/com/scalar/db/storage/cosmos/CosmosOperationChecker.java

```diff
@@ -2,36 +2,130 @@
 
 import com.scalar.db.api.ConditionalExpression;
 import com.scalar.db.api.Delete;
+import com.scalar.db.api.Get;
 import com.scalar.db.api.Mutation;
+import com.scalar.db.api.Operation;
 import com.scalar.db.api.Put;
+import com.scalar.db.api.Scan;
 import com.scalar.db.api.TableMetadata;
 import com.scalar.db.common.TableMetadataManager;
 import com.scalar.db.common.checker.OperationChecker;
 import com.scalar.db.common.error.CoreError;
 import com.scalar.db.config.DatabaseConfig;
 import com.scalar.db.exception.storage.ExecutionException;
+import com.scalar.db.io.BigIntColumn;
+import com.scalar.db.io.BlobColumn;
+import com.scalar.db.io.BooleanColumn;
+import com.scalar.db.io.ColumnVisitor;
 import com.scalar.db.io.DataType;
+import com.scalar.db.io.DoubleColumn;
+import com.scalar.db.io.FloatColumn;
+import com.scalar.db.io.IntColumn;
+import com.scalar.db.io.TextColumn;
 
 public class CosmosOperationChecker extends OperationChecker {
+
+  private static final char[] ILLEGAL_CHARACTERS_IN_PRIMARY_KEY = {
+    // Colons are not allowed in primary key columns due to the `ConcatenationVisitor` limitation
+    ':',
+
+    // The following characters are not allowed in primary key columns because they are restricted
+    // and cannot be used in the `Id` property of a Cosmos DB document. See the following link for
+    // more information:
+    // https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.databaseproperties.id?view=azure-dotnet#remarks
+    '/',
+    '\\',
```

Comment:
```suggestion
    '\',
```

The [document](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.resource.id?view=azure-dotnet#remarks) states the character `\\` (double backslashes) is restricted.

However, it seems that we cannot use the character `\` (single backslash) too.

By using ScalarDB 3.13.1, I created the test table and inserted test records that include `\` and `\\`.

```sql
0: scalardb> create table foo (a text primary key, b int);
No rows affected (0.508 seconds)
0: scalardb>
0: scalardb> insert into foo values ('aaa\bbb', 0);
1 row affected (0.27 seconds)
0: scalardb>
0: scalardb> insert into foo values ('aaa\\bbb', 0);
1 row affected (0.102 seconds)
```

After that, I retrieved both records respectively from the Azure Portal.

As a result, both `aaa\bbb` and `aaa\\bbb` return the following errors.

- `\` (single backslash)
  ```console
  Failed to read item aaa\bbb:
   Illegal characters ['/', '\', '#'] cannot be used in Resource ID
  ```
- `\\` (double backslashes)
  ```console
  Failed to read item aaa\\bbb:
   Illegal characters ['/', '\', '#'] cannot be used in Resource ID
  ```

So, I think the document description `\\` is a typo...
(I guess, they may want to represent the single `\`, by using `\` as an escape character, but it seems that it does not work properly. As a result, maybe, the document states `\\`...)

Therefore, I think it would be better to validate `\` (single backslash) instead of `\\` (double backslashes).

Reply:
`\\` represents a single backslash (the first backslash is for escaping). So, the current code should be correct. Or, when you ran this fix, were backslashes not validated correctly?

Reply:
Thank you for your reply!

> `\\` represents a single backslash (the first backslash is for escaping) in Java (and probably C#). So, the current code should be correct. 

Ah, I see... Sorry, I thought that `\\` does not mean `\` in the validation process when I read the code...

> Or, when you ran this fix, were backslashes not validated correctly?

I just noticed it in the code review before I tried the actual behavior, so I have not tested it yet.

I will test the validation behavior from now on. So, please give me some time.

Reply:
I was able to confirm that all illegal characters (including `\`) are restricted on the ScalarDB side in my local environment as follows!

<details>
<summary> Test result</summary>

- Create a test table
  ```sql
  0: scalardb> create table foo (a text, b text, c int, primary key (a,b));
  No rows affected (0.553 seconds)
  0: scalardb>
  0: scalardb> describe foo;
  +------------+------+--------------+----------------+-----------------+-----------------+-----------+-------------+
  | columnName | type | isPrimaryKey | isPartitionKey | isClusteringKey | clusteringOrder | isIndexed | isEncrypted |
  +------------+------+--------------+----------------+-----------------+-----------------+-----------+-------------+
  | a          | TEXT | true         | true           | false           | null            | false     | false       |
  | b          | TEXT | true         | false          | true            | ASC             | false     | false       |
  | c          | INT  | false        | false          | false           | null            | false     | false       |
  +------------+------+--------------+----------------+-----------------+-----------------+-----------+-------------+
  3 rows selected (0.043 seconds)
  ```
- INSERT
  - Partition key
    ```sql
    0: scalardb> insert into foo values ('aaa/bbb', 'b', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa/bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('aaa\bbb', 'b', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa\bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('aaa#bbb', 'b', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa#bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('aaa?bbb', 'b', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa?bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('aaa:bbb', 'b', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa:bbb) (state=42000,code=501)
    ```
  - Clustering key
    ```sql
    0: scalardb> insert into foo values ('a', 'xxx/yyy', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx/yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('a', 'xxx\yyy', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx\yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('a', 'xxx#yyy', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx#yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('a', 'xxx?yyy', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx?yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> insert into foo values ('a', 'xxx:yyy', 0);
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx:yyy) (state=42000,code=501)
    ```
- SELECT
  - Partition key
    ```sql
    0: scalardb> select * from foo where a = 'aaa/bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa/bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'aaa\bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa\bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'aaa#bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa#bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'aaa?bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa?bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'aaa:bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa:bbb) (state=42000,code=501)
    ```
  - Clustering key
    ```sql
    0: scalardb> select * from foo where a = 'a' and b = 'xxx/yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx/yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'a' and b = 'xxx\yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx\yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'a' and b = 'xxx#yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx#yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'a' and b = 'xxx?yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx?yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> select * from foo where a = 'a' and b = 'xxx:yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx:yyy) (state=42000,code=501)
    ```
- UPDATE
  - Partition key
    ```sql
    0: scalardb> update foo set c = 100 where a = 'aaa/bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa/bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'aaa\bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa\bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'aaa#bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa#bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'aaa?bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa?bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'aaa:bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa:bbb) (state=42000,code=501)
    ```
  - Clustering key
    ```sql
    0: scalardb> update foo set c = 100 where a = 'a' and b = 'xxx/yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx/yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'a' and b = 'xxx\yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx\yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'a' and b = 'xxx#yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx#yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'a' and b = 'xxx?yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx?yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> update foo set c = 100 where a = 'a' and b = 'xxx:yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx:yyy) (state=42000,code=501)
    ```
- DELETE
  - Partition key
    ```sql
    0: scalardb> delete from foo where a = 'aaa/bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa/bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'aaa\bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa\bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'aaa#bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa#bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'aaa?bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa?bbb) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'aaa:bbb';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column a in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: aaa:bbb) (state=42000,code=501)
    ```
  - Clustering key
    ```sql
    0: scalardb> delete from foo where a = 'a' and b = 'xxx/yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx/yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'a' and b = 'xxx\yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx\yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'a' and b = 'xxx#yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx#yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'a' and b = 'xxx?yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx?yyy) (state=42000,code=501)
    0: scalardb>
    0: scalardb> delete from foo where a = 'a' and b = 'xxx:yyy';
    Error: Invalid query (INVALID_ARGUMENT: CORE-10145: The value of the column b in the primary key contains an illegal character. Primary-key columns must not contain any of the following characters: ':', '/', '\', '#', '?'. Value: xxx:yyy) (state=42000,code=501)
    ```
</details>

Reply:
@kota2and3kan Thank you for the confirmation! 🙇